kitsune/wiki/utils.py

```diff
@@ -290,3 +296,70 @@ def build_topics_data(request: HttpRequest, product: Product, topics: list[Topic
         topics_data.append(topic_data)
 
     return topics_data
+
+
+def clean_kb_visited(session: SessionBase, ttl: int = KB_VISITED_DEFAULT_TTL) -> None:
+    """
+    Remove expired slugs from the "kb-visited" dictionary within the given session
+    based on the given TTL.
+    """
+    if not (session and ("kb-visited" in session)):
+        return
+
+    now = time.time()
+    kb_visited = session["kb-visited"]
+
+    # Remove any slugs that have expired.
+    if expired_slugs := [slug for slug, ts in kb_visited.items() if (now - ts) > ttl]:
+        for slug in expired_slugs:
+            del kb_visited[slug]
+        session.modified = True
+
+
+def update_kb_visited(
+    session: SessionBase, doc: Document, ttl: int = KB_VISITED_DEFAULT_TTL
+) -> None:
+    """
+    Cleans and then updates the "kb-visited" dictionary within the given session.
+    """
+    if not session:
+        return
+
+    if "kb-visited" in session:
+        clean_kb_visited(session, ttl=ttl)
+        kb_visited = session["kb-visited"]
+    else:
+        kb_visited = session["kb-visited"] = {}
+
+    # Note that we have visited the given KB article.
+    kb_visited[doc.original.slug] = time.time()
```

Comment:
suggestion: We could have the kb_visited dictionary to hold list of slugs per topic.

`kb_visited {'topic' : [slug1, slug2]`

Reply:
I've modified this considerably with something in the same vein as you're suggesting (but I had to include the products as well).